### PR TITLE
chore: add retry to tests requiring forwarding

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
# About this change - What it does

References: #xxxxx

# Why this way

### CODE REVIEW

File: tests/integration/utils/rest_client.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+class UnexpectedResponseStatus(Exception):
+    pass
+
+
+class RetryRestClient:
+    def __init__(self, client: Client):
+        self._karapace_client = client
+
+    @retry(stop=stop_after_attempt(5))
```
Comment (keejon) [inline tests/integration/utils/rest_client.py:24]: Should we maybe wait a bit between retries?

File: tests/integration/utils/rest_client.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+Copyright (c) 2024 Aiven Ltd
+See LICENSE for details
+"""
+from __future__ import annotations
+
+from aiohttp import BasicAuth
+from collections.abc import Mapping
+from karapace.client import Client, Headers, Path, Result
+from karapace.typing import JsonData
+from tenacity import retry, stop_after_attempt
```
Comment (nosahama) [inline tests/integration/utils/rest_client.py:13]: Noice 💯

File: tests/integration/test_schema_registry_auth.py [hunk 3]
```diff
@@ -29,139 +29,157 @@
...
     )
     assert res.status_code == 200
     sc_id = res.json()["id"]
     assert sc_id >= 0
 
-    res = await registry_async_client_auth.get(f"subjects/{quote(subject)}/versions/latest")
+    res = await registry_async_retry_client_auth.get(
+        f"subjects/{quote(subject)}/versions/latest", expected_response_code=401
+    )
     assert res.status_code == 401
```
Comment (nosahama) [inline tests/integration/test_schema_registry_auth.py:50]: **[non-blocking]**
I know it'd be a harder refactor, but we also somehow already assert the response code:
```
if response.status_code != expected_response_code:
    raise UnexpectedResponseStatus(f"Unexpected status code: {response!r}")
```
we can omit the assertion `assert res.status_code == 401` in all places where we use the `retry_client` or even move the assertion into the client stub.

File: tests/integration/utils/rest_client.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+from aiohttp import BasicAuth
+from collections.abc import Mapping
+from karapace.client import Client, Headers, Path, Result
+from karapace.typing import JsonData
+from tenacity import retry, stop_after_attempt, wait_fixed
+from typing import Final
+
+
+RETRY_WAIT: Final = 0.5
```
Comment (nosahama) [inline tests/integration/utils/rest_client.py:17]: nanoseconds 😄

**Review by nosahama — DISMISSED** [review]

File: tests/integration/test_schema_registry_auth.py [hunk 5]
```diff
@@ -198,8 +218,7 @@ async def test_sr_auth_forwarding(registry_async_auth_pair: List[str]) -> None:
         while True:
             assert counter < max_tries, "Compat update not propagated"
             resp = requests.get(f"{primary_url}/config", auth=auth)
-            if not resp.ok:
-                continue
+            assert resp.ok
```
Comment (keejon) [inline tests/integration/test_schema_registry_auth.py:221]: Is this change intentional? It will fail now instead of retrying.

Comment (nosahama) [reply tests/integration/test_schema_registry_auth.py:221]: As we have a retry alreadt, maybe no need to continue, but then we have this check above `assert counter < max_tries, "Compat update not propagated"`, not sure if it's used for the old retry style.

Comment (jjaakola-aiven) [reply tests/integration/test_schema_registry_auth.py:221]: No, good catch.

File: tests/integration/utils/rest_client.py (hunk 1 above)
Comment (jjaakola-aiven) [reply tests/integration/utils/rest_client.py]: Done.

File: tests/integration/utils/rest_client.py (hunk 4 above)
Comment (jjaakola-aiven) [reply tests/integration/utils/rest_client.py]: Renamed.

File: tests/integration/test_schema_registry_auth.py (hunk 5 above)
Comment (jjaakola-aiven) [reply tests/integration/test_schema_registry_auth.py]: Done.

File: tests/integration/test_schema_registry_auth.py (hunk 3 above)
Comment (jjaakola-aiven) [reply tests/integration/test_schema_registry_auth.py:50]: Requires bit more work. I'd like to have this in first as these do reduce the flakiness.

**Review by keejon — DISMISSED** [review]
LGTM 👍

File: tests/integration/test_schema_registry_auth.py [hunk 6]
```diff
@@ -170,34 +188,36 @@ async def test_sr_ids(registry_async_client_auth: Client) -> None:
...
+async def test_sr_auth_forwarding(
+    registry_async_auth_pair: List[str], registry_async_retry_client_auth: RetryRestClient
+) -> None:
     # Test primary/replica forwarding with global config setting
     primary_url, replica_url = registry_async_auth_pair
     max_tries, counter = 5, 0
     wait_time = 0.5
     for compat in ["FULL", "BACKWARD", "FORWARD", "NONE"]:
-        resp = requests.put(f"{replica_url}/config", json={"compatibility": compat}, auth=auth)
+        resp = await registry_async_retry_client_auth.put(f"{replica_url}/config", json={"compatibility": compat}, auth=admin)
```
Comment (nosahama) [inline tests/integration/test_schema_registry_auth.py:216]: This was failing and we had to split the used `auth`, `aiohttp` requires the `BasicAuth` and `requests` requires something different, but we were first passing the requests auth to the aiohttp client and it was failing.

**Review by AnatolyPopov — COMMENTED** [review]
LGTM
